WordPress/gutenberg
# Custom Link: Decode value in URL input field

Fixes #55537
Related to #46435

## What?

This PR fixes an issue where the URL cannot be changed correctly from the navigation sidebar in the Custom Link block.

| Before| After |
|--------|--------|
| ![before](https://github.com/WordPress/gutenberg/assets/54422211/124985c0-c111-413a-88c4-909229b9a948) | ![after](https://github.com/WordPress/gutenberg/assets/54422211/6edecfeb-5c79-4c12-b980-a5d900606f8f) | 

## Why?

The URL in this block is sanitized by `esc_url()` when rendered. As a result, the correct URL may not be rendered, for example if it contains parameters as shown below.

```php
echo esc_url( 'https://google.com?param=<%>' );

// > https://google.com?param=%
````

My understanding is that to prevent this, it was changed in #19679 to keep the encoded URL as an attribute value. [This comment](https://github.com/WordPress/gutenberg/pull/46435/files#r1062236863) explains more about why. And then #46435 was filed to properly decode the encoded URL on the front end.

On the other hand, although the URL can be changed from the block sidebar, the encoded URL was displayed as the value of the text field. The encoded URL may represent one character with multiple characters. Deleting one of those characters by keyboard will corrupt the encoded string, and the block will internally attempt to repair the URL, causing problems like the one reported in #55537 to appear.

User-editable fields should display the decoded URL instead of the encoded URL.

## How?

In the URL input field, the encoded URL was decoded using the `safeDecodeURI()` function.

## Testing Instructions

- Click the inserter in the navigation block. Enter `https://google.com?param=<%>` in the text field.
- A Custom Link block should now be created.
- Open the block sidebar.
  - trunk: The URL field should display the encoded URL (`https://google.com?param=%3C%25%3E`).
  - This PR: The URL field should display the decoded URL (`https://google.com?param=<%>`).
- Try editing the URL value, perhaps using the delete key.
  - trunk: You may not be able to delete characters, or unintended encoded characters may be added.
  - This PR: You should be able to change all characters normally.
-  Check the link for this block in the front end.
- The URL should show the correct `https://google.com?param=<%>` without being stripped.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/54422211/ab1bf253-052c-478c-9e42-77d4154d6a04
